### DISCUSSION TITLE
Ctrl A To Select The Whole Buffer

### DISCUSSION THREAD

**Discussion by blueray453** [fish-shell/fish-shell#9829] [Q&A] [answered]
I want to make a keybinding which will have the following sequence

```
function ctrl_a
    beginning-of-buffer
    begin-selection
    end-of-buffer
    end-selection
end

```
I understand that I have to make a function and link it like

```
function fish_user_key_bindings
    bind \ca ctrl_a
end
```

My question is, where to put the function and what is the standard way of doing it?

The second issue is, When I use ctrl+c the terminal looks like:

```
%
ls^C
```
Hopefully you see the `^C`. I do not want that. What to do?

The last question is, how to unbind a key? Suppose, I want to unbind ctrl+c, what to do?

**Comment by faho** [comment, marked as answer]
>I want to make a keybinding which will have the following sequence

These are all "special input functions", so you can just use them like

```fish
bind \ca beginning-of-buffer begin-selection end-of-buffer end-selection
```

Not that it'll *do* anything, because `end-selection` will, well, end the selection. It will stop selecting things.

So you want

```fish
bind \ca beginning-of-buffer begin-selection end-of-buffer
```

>I understand that I have to make a function and link it like

You do not have to use `fish_user_key_bindings`. That's outdated. You can just put your `bind ...` into config.fish.

>My question is, where to put the function and what is the standard way of doing it?

Once more, the standard way to save a function is to save it in ~/.config/fish/functions. This is what `funcsave` will do, for instance.

You can also just put it in config.fish, and it's fine.

In this case, I recommend not bothering with the function because that's just unnecessary indirection.

In particular, once you run a function, you need to run "special input functions" via `commandline -f`, so it would have to be

```fish
function ctrl_a
    commandline -f beginning-of-buffer
    commandline -f begin-selection
    commandline -f end-of-buffer
end
```

>The last question is, how to unbind a key? Suppose, I want to unbind ctrl+c, what to do?

Typically, you would just rebind it. If you really want it to not do anything, just bind it to `true` or `false`:

```fish
bind \cc true
```

**Comment by blueray453** [comment]
Why does it show `^C` in terminal when I press ctrl+C

```
%
ls^C
%
```
I do no want that.

And is there any way to unbind all keybindings (using a loop)?

**Reply by faho** [reply]
> Why does it show ^C in terminal when I press ctrl+C

Because that's what the default ctrl+c binding does. You haven't rebound it.

> And is there any way to unbind all keybindings (using a loop)?

```fish
bind --erase --all --preset
```

You need to run this *after* the bindings have been set up, which happens *after* config.fish.

Most likely what you want to do is make your own binding function, and then set $fish_key_bindings to its name, like

```fish
function blueray_bindings
    bind --erase --all --preset # <- removes all bindings
    bind --preset '' self-insert # <- the binding for inserting normal characters
    bind --preset \r execute
    bind --preset -k backspace backward-delete-char
end

set -g fish_key_bindings blueray_bindings
```

The function can be in an autoload file, the `set` should be in config.fish.

**Comment by blueray453** [comment]
If I create the `ismail_bindings` function in:

/home/ismail/.config/fish/functions/ismail_bindings.fish

Then will it be auto-loaded?

I mean there is a funcsave function (it says `funcsave: wrote /home/ismail/.config/fish/functions/ismail_bindings.fish`). Not understanding how to actually autoload a function.

I have checked both

https://fishshell.com/docs/current/cmds/funcsave.html

and 

https://fishshell.com/docs/current/language.html#syntax-function-autoloading

but it is not clear to me how to autoload a function.

**Reply by faho** [reply]
>but it is not clear to me how to autoload a function.

*Fish* autoloads a function.

When you tell fish to run "foo", it looks if there is a file "foo.fish" in ~/.config/fish/functions/. If there is, it loads that file. If it then has a function called "foo", it runs it.

**Comment by blueray453** [comment]
One more thing, have you checked `bind \ca beginning-of-buffer begin-selection end-of-buffer`. It does not do anything in my machine except moving cursor. It does not select (I am assuming, selected text will be **highlighted**).

I have also tested with both:

    bind \ca beginning-of-buffer begin-selection end-of-buffer togglecase-selection

and

    bind \ca beginning-of-buffer begin-selection end-of-buffer end-selection togglecase-selection

It does not change case, it means at least to me that it did not actually select.

**Reply by faho** [reply]
>t does not do anything in my machine except moving cursor. It does not select (I am assuming, selected text will be highlighted).

Please read [my comment](https://github.com/fish-shell/fish-shell/discussions/9829#discussioncomment-6075124), where I told you just that.


>Not that it'll do anything, because end-selection will, well, end the selection. It will stop selecting things.

>So you want

>bind \ca beginning-of-buffer begin-selection end-of-buffer

**Reply by blueray453** [reply]
As i said in my last comment, I first tried `bind \ca beginning-of-buffer begin-selection end-of-buffer`. It does not do anything. So I tried the other commands later. Even `bind \ca beginning-of-buffer begin-selection end-of-buffer togglecase-selection` does not do anything.

**Reply by blueray453** [reply]
But 
```
function ctrl_a
    commandline -f beginning-of-buffer
    commandline -f begin-selection
    commandline -f end-of-buffer
end
```

and

    bind \ca ctrl_a

works.

**Reply by faho** [reply]
Both of these work for me:

```fish
bind \ca beginning-of-buffer begin-selection end-of-buffer
bind \ca beginning-of-buffer begin-selection end-of-buffer togglecase-selection
```

the former selects all the text in the commandline, which is colored bold white text on a grey background in the default theme.

The latter additionally changes it to SHOUTY CAPS, and if you run it again to "quiet smalls".

**Reply by blueray453** [reply]
The selection currently looks like:

![current-selection](https://github.com/fish-shell/fish-shell/assets/855297/c8bf5015-ec66-41fb-b7fe-98482ce7384c)

There is no highlighting.

How to make it look like:

![selection](https://github.com/fish-shell/fish-shell/assets/855297/76a70b21-1f5c-476d-a39d-de0aaf31c21e)

otherwise it is very hard to understand.

And I am very sorry for bothering you this much. Hopefully in 2/3 days, I will be all set and will not disturb you this frequently (maybe once in 2/3 months).

Anyways, You probably missed 

https://github.com/fish-shell/fish-shell/discussions/9829#discussioncomment-6075171

and 

https://github.com/fish-shell/fish-shell/discussions/9829#discussioncomment-6075350

**Reply by faho** [reply]
> The selection currently looks like:

This can mean any of a few things:

1. You don't have a selection - possibly because you made a mistake with the binding
2. You don't have a color terminal (or at least fish doesn't think you do)
3. You haven't set a selection color

You can check these. Run `echo normal(set_color $fish_color_selection)selection(set_color normal)normal`. If this prints something in a selection color then 2 & 3 aren't it.

Then you'll want to run `bind \ca`, to see what that is actually currently bound to.

**Reply by blueray453** [reply]
So the output of your command is. I am using 'Solarized Dark' theme.

    echo normal(set_color $fish_color_selection)selection(set_color normal)normal
    set_color: Unknown color 'white --bold --background=brblack'

I am using Alacritty terminal.

I saw your comment:

> Because that's what the default ctrl+c binding does. You haven't rebound it.

My issue is with the marker `^C`. I do not want that.

I tried 

    bind \cc cancel

Instead of 

    bind \cc cancel-commandline

Because the documentation says:

>cancel
>
>    cancel the current commandline and replace it with a new empty one
>
>cancel-commandline
>
>    cancel the current commandline and replace it with a new empty one, leaving the old one in place with a marker to show that it was cancelled

However, it does not give me the expected behavior. 

Moreover, I bind `cancel-commandline` with ctrl+q , still it shows ^C.

**Reply by faho** [reply]
> set_color: Unknown color 'white --bold --background=brblack'

`set --show fish_color_selection` (no `$`) says?

> My issue is with the marker ^C. I do not want that.

Okay, try

```fish
bind \cc 'commandline -r ""'
```

> Moreover, I bind cancel-commandline with ctrl+q , still it shows ^C.

Yes, cancel-commandline prints `^C`. Slightly naughty, but tbh barely anyone tries binding it to anything else.

**Comment by blueray453** [comment]
Here is the output:

```
% set --show fish_color_selection
$fish_color_selection: set in universal scope, unexported, with 1 elements
$fish_color_selection[1]: |white --bold --background=brblack|
```

**Reply by faho** [reply]
Well, there's your problem.

This variable was set like

```fish
set --universal fish_color_selection "white --bold --background=brblack"
```

when it should have been

```fish
set --universal fish_color_selection white --bold --background=brblack
```

I just checked and neither the web config tool nor `fish_config theme choose` act like that in fish 3.6.1.

**Reply by blueray453** [reply]
I tried 

    bind \cc 'commandline -r ""'

This is not what I wanted. It just clear the prompt.

I want to overwrite `/usr/share/fish/functions/__fish_cancel_commandline.fish`

I copy pasted it in `/home/ismail/.config/fish/functions`

Replaced `set -g __fish_cancel_text "^C"` with `set -g __fish_cancel_text ""`, does not work.

I directly edited `/usr/share/fish/functions/__fish_cancel_commandline.fish` with sudo (replaced that line). Does not work. Maybe this function is cached somewhere.

How to overwrite this function? Or is there any simpler solution for this?

**Reply by faho** [reply]
>Maybe this function is cached somewhere.

https://fishshell.com/docs/current/language.html#autoloading-functions.

It is cached *inside the running shell*, in memory.

This is why I mentioned `funced` and `funcsave`.

**Reply by blueray453** [reply]
So, how do i edit `/usr/share/fish/functions/__fish_cancel_commandline.fish` with `funced` and then save with `funcsave`.

I have tried:

```
funced __fish_cancel_commandline
funced __fish_setup_cancel_text
```

It create new functions. It does not open the file I am looking for.

**Reply by ridiculousfish** [reply]
Does `funced --save __fish_cancel_commandline` do what you want?

**Reply by ridiculousfish** [reply]
Note that `__fish_cancel_commandline` is no longer invoked by default (see [this commit](https://github.com/fish-shell/fish-shell/commit/eb35975c0fdd7420608aa743882c0fc96f3bf585), not sure if we documented it as it's internal). If you want it to run, you will want to bind it, e.g. via `bind \cc __fish_cancel_commandline` (or just make your own function).

**Reply by blueray453** [reply]
@ridiculousfish The commands that you showed works. But editing those functions does not work.

The commits you linked has a comment like:

> leaving the old one in place with a marker to show that it was cancelled

I want to:

1. leaving the old one in place
2. change this marker to '^Q' 

Editing `__fish_cancel_commandline` actually does not help. It has the same effect of `bind \cc 'commandline -r ""'`. It does not "leaving the old one in place".

Which file shall I overwrite to change the marker of `cancel_commandline`?

Or, if it is difficult then you can let us pass an argument to `cancel_commandline` to change the marker. It will also help.
